### PR TITLE
Patch/11 support client request timeout

### LINKED ISSUE #11: Add client config `request_timeout`
I saw there's a `request_timeout` on the server config, that makes sense.

However, I think it can be useful to also set a client config `request_timeout`. So basically, if the client on is set, then take that as timeout, or else default to the server config's `request_timeout`.

Does that make sense to add to this lib?

### PR DESCRIPTION
**Pull request by GlenDC**
Here you go. this resolves #11.
Resolved it as a thank you for making this library, it fulfils my direct needs,
and shows production value.

Also, I do want a request timeout myself on client side, as part of my proxy-gateway project.

There are no unit tests, so cannot run those. But all your three examples do run.

### CODE REVIEW

File: src/util/stream.rs [hunk 1]
```diff
@@ -38,3 +44,37 @@ macro_rules! ready {
...
     };
 }
+
+pub async fn tcp_connect_with_timeout<T>(addr: T, request_timeout_s: u64) -> Result<TcpStream>
+    where T: ToSocketAddrs,
+{
+    let fut = tcp_connect(addr);
+    match timeout(Duration::from_secs(request_timeout_s), fut).await {
+        Ok(result) => result,
+        Err(_) => Err(ReplyError::TtlExpired.into()),
```
Comment (dizda) [inline src/util/stream.rs:54]: I wonder if we should create a new kind of error, eg. `ConnectTimeout`? What do you think?

File: src/client.rs [hunk 2]
```diff
@@ -18,18 +19,26 @@ const MAX_ADDR_LEN: usize = 260;
 
 #[derive(Debug)]
 pub struct Config {
+    /// Timeout of the command request
+    request_timeout: Option<u64>,
```
Comment (dizda) [inline src/client.rs:23]: This is not a timeout related to request, but to the connection.

`connect_timeout` would be more preferable imo, what do you think?

File: src/util/stream.rs (hunk 1 above)
Comment (GlenDC) [reply src/util/stream.rs:54]: Makes sense to me. Going to add it.

File: src/client.rs (hunk 2 above)
Comment (GlenDC) [reply src/client.rs:23]: I do agree, but unless I misunderstood, it seemed to me the same for the server functionality?
I also wonder if we also need another timeout for sending content as well as receiving content. As that is currently not something we (can) configure either, right?

Comment (GlenDC) [reply src/client.rs]: Changed it for the client.
I do feel we ought to change it for the server as well,
as it is used for the exact same purpose. But that would be a breaking change, so didn't touch it yet.

File: src/util/stream.rs (hunk 1 above)
Comment (GlenDC) [reply src/util/stream.rs]: Added it. There doesn't seem to be a specific Socks5 error we can make to (going from the spec), so I still map it to TTL expired.

**Comment by Yneth** [comment]
could you please add the same for socks4 client?

**Comment by GlenDC** [comment]
There was a request to also support this feature in socks4 client, but don't think it has a config. So going to leave that request without further action.

**Comment by dizda** [comment]
Thank you @GlenDC 🙏

**Pull request merged**